doc/kernel/usermode/kernelobjects.rst

```diff
@@ -0,0 +1,226 @@
+.. _kernelobjects:
+
+Kernel Objects
+##############
+
+A kernel object can be one of three classes of data:
+
+* A core kernel object, such as a semaphore, thread, pipe, etc.
+* A thread stack, which is an array of :c:type:`struct _k_thread_stack_element`
+  and declared with :c:macro:`K_THREAD_STACK_DEFINE()`
+* A device driver instance (struct device) that belongs to one of a defined
+  set of subsystems
+
+The set of known kernel objects and driver subsystems is defined in
+include/kernel.h as :cpp:enum:`k_objects`.
+
+Kernel objects are completely opaque to user threads. User threads work
+with addresses to kernel objects when making API calls, but may never
+dereference these addresses, doing so will cause a memory protection fault.
+All kernel objects must be placed in memory that is not acessible by
+user threads.
+
+Since user threads may not directly manipulate kernel objects, all use of
+them must go through system calls. In order to perform a system call on
+a kernel object, checks are performed by system call handler functions
+that the kernel object address is valid and that the calling thread
+has sufficient permissions to work with it.
+
+Object Placement
+================
+
+Kernel objects that are only used by supervisor threads have no restrictions.
+In order for a kernel object to be usable by a user thread, several conditions
+must be met on how it is declared.
+
+First, that object must be declared as a top-level global at build time, such
+that it appears in the ELF symbol table. It is permitted to declare kernel
+objects with static scope. The post-build script ``gen_kobject_list.py`` scans
+the generated ELF file to find kernel objects and places their memory addresses
+in a special table of kernel object metadata.  Kernel objects may be members of
+arrays or embedded within other data structures.
+
+Kernel objects must be located in memory reserved for the kernel. If
+:option:`CONFIG_APPLICATION_MEMORY` is used, all declarations of kernel objects
+inside application code must be prefixed with the :c:macro:`__kernel` attribute
+so that they are placed in the right memory sections. The APIs for statically
+declaring and initializing kernel objects (such as :c:macro:`K_SEM_DEFINE()`)
+automatically do this. However, uninitialized kernel objects need to be tagged
+like this:
+
+.. code-block:: c
+
+    __kernel struct k_sem my_sem;
+
+Any memory reserved for a kernel object must be used exclusively for that
+object. Kernel objects may not be members of a union data type.
```

Comment:
Is there a way to catch this during build time? is it worth it?

Reply:
yes, build time will catch if the kernel object is part of a union and not add it to the gperf table.